flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/asyncprocessing/AbstractAsyncStateStreamOperator.java

```diff
@@ -0,0 +1,134 @@
+        implements AsyncStateProcessing {
+
+    private AsyncExecutionController asyncExecutionController;
+
+    private RecordContext lastProcessContext;
+
+    /** Initialize necessary state components for {@link AbstractStreamOperator}. */
+    @Override
+    public void setup(
+            StreamTask<?, ?> containingTask,
+            StreamConfig config,
+            Output<StreamRecord<OUT>> output) {
+        super.setup(containingTask, config, output);
+        // TODO: properly read config and setup
+        final MailboxExecutor mailboxExecutor =
+                containingTask.getEnvironment().getMainMailboxExecutor();
+        this.asyncExecutionController = new AsyncExecutionController(mailboxExecutor, null);
+    }
+
+    @Override
+    public final boolean isAsyncStateProcessingEnabled() {
+        // TODO: Read from config
+        return true;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public final <T> void setAsyncKeyedContextElement(
+            StreamRecord<T> record, KeySelector<T, ?> keySelector) throws Exception {
+        lastProcessContext =
+                asyncExecutionController.buildContext(
+                        record.getValue(), keySelector.getKey(record.getValue()));
+        // The processElement will be treated as a callback for dummy request. So reference
+        // counting should be maintained.
+        // When state request submitted, ref count +1, as described in FLIP-425:
+        // To cover the statements without a callback, in addition to the reference count marked
+        // in Fig.5, each state request itself is also protected by a paired reference count.
+        lastProcessContext.retain();
+        asyncExecutionController.setCurrentContext(lastProcessContext);
+    }
+
+    @Override
+    public final void postProcessElement() {
+        // The processElement will be treated as a callback for dummy request. So reference
+        // counting should be maintained.
+        // When a state request completes, ref count -1, as described in FLIP-425:
+        // To cover the statements without a callback, in addition to the reference count marked
+        // in Fig.5, each state request itself is also protected by a paired reference count.
+        lastProcessContext.release();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public final <T> ThrowingConsumer<StreamRecord<T>, Exception> getRecordProcessor(int inputId) {
+        // Ideally, only TwoStreamInputOperator/OneInputStreamOperator(Input) will invoke here.
+        // Only those operators have the definition of processElement(1/2).
+        if (this instanceof TwoInputStreamOperator) {
+            switch (inputId) {
+                case 1:
+                    return record -> {
+                        setAsyncKeyedContextElement(record, (KeySelector<T, ?>) stateKeySelector1);
+                        ((TwoInputStreamOperator) this).processElement1(record);
```

Comment:
Should the `setKeyContextElement1()`  also be invoked here?

Reply:
Afraid not. I intended to keep `setKeyContextElement1` as only for the synchronous execution.